data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/sql/DefaultSqlPreparedQuery.java

```diff
@@ -149,6 +149,12 @@ public void bindParameters(Binder binder, InvocationContext<?, ?> invocationCont
     private int getQueryParameterValueSize(QueryParameterBinding parameter) {
         int parameterIndex = parameter.getParameterIndex();
         if (parameterIndex == -1) {
+            // If parameter (in criteria for example) is multivalued then actual parameter value size is
+            // the size of that Collection, Iterable or array.
+            // Issue https://github.com/micronaut-projects/micronaut-data/issues/1678
+            if (isMultivaluedParameter(parameter.getValue())) {
```

Comment:
I think the original assumption that values not extracted from the parameters don't need expansion was false. I don't think there is a need to check if the value is multi-valued, just write something like:
```java
Object value;
if (parameterIndex == -1) {
 value = parameter.getValue();
} else { 
 value = preparedQuery.getParameterArray()[parameterIndex]
 }
return sizeOf(value);

Reply:
I will try that, tnx.

Reply:
Updated, fix is still ok as test passes. Thanks